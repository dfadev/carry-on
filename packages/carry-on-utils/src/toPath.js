// adapt lodash toPath

/** Used to match property names within property paths. */
const reLeadingDot = /^\./;
const rePropName =
  /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g;

/** Used to match backslashes in property paths. */
const reEscapeChar = /\\(\\)?/g;

export default function toPath(string) {
  const result = [];
  if (reLeadingDot.test(string)) {
    result.push("");
  }
  string.replace(rePropName, (match, number, quote, str) => {
    result.push(quote ? str.replace(reEscapeChar, "$1") : number || match);
  });

  return result;
}
