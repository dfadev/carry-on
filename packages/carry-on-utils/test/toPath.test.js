import toPath from "../src/toPath";

test("should handle a leading dot", () => {
  const result = toPath(".field.path");
  expect(result).toMatchSnapshot();
});

test("should handle quoted keys", () => {
  const result = toPath('field["quoted.key"].path');
  expect(result).toMatchSnapshot();
});

test("should handle undefined", () => {
  const undefinedResult = toPath(undefined);
  expect(undefinedResult).toMatchSnapshot();
});

test("should handle null", () => {
  const nullResult = toPath(null);
  expect(nullResult).toMatchSnapshot();
});
