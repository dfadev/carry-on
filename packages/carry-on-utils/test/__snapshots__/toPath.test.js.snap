// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle a leading dot 1`] = `
Array [
  "",
  "field",
  "path",
]
`;

exports[`should handle quoted keys 1`] = `
Array [
  "field",
  "quoted.key",
  "path",
]
`;
